pkg/crawler/crawler.go

```diff
@@ -262,3 +308,204 @@ func (c *Crawler) fetchSHA1(url string) ([]byte, error) {
 }
+
+func (c *Crawler) fetchAndSavePOMLicenseKeys(url string) ([]string, error) {
+	var keys []string
+	resp, err := c.http.Get(url)
+	if resp.StatusCode == http.StatusNotFound {
+		return keys, nil
+	}
+	if err != nil {
+		return keys, xerrors.Errorf("can't get pom xml from %s: %w", url, err)
+	}
+	defer resp.Body.Close()
+
+	var pomProject PomProject
+
+	decoder := xml.NewDecoder(resp.Body)
+	decoder.CharsetReader = charset.NewReaderLabel
+	err = decoder.Decode(&pomProject)
+
+	if err != nil {
+		return keys, xerrors.Errorf("can't parse pom xml from %s: %w", url, err)
+	}
+
+	if len(pomProject.Licenses) == 0 {
+		return keys, nil
+	}
+
+	for _, l := range pomProject.Licenses {
+		l.LicenseKey = getLicenseKey(l)
+
+		// update uniqueLicenseKeys map
+		c.uniqueLicenseKeys.Set(l.LicenseKey, l)
+
+		keys = append(keys, l.LicenseKey)
+	}
+
+	return keys, nil
+
+}
+
+func (c *Crawler) classifyLicense() error {
+	normalizedLicenseMap := make(map[string]string)
+
+	// prepare classifier data i.e create temporary files with license text to be used for classification
+	filesLicenseMap := c.prepareClassifierData()
+
+	if len(filesLicenseMap.Keys()) == 0 {
+		return nil
+	}
+
+	// classify licenses
+
+	// 1 minute is the timeout for license classification of a file
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	// 1000 is the number of concurrent tasks spawned to process license files
+	errs := c.classifier.ClassifyLicensesWithContext(ctx, 1000, filesLicenseMap.Keys(), true)
+	if len(errs) > 0 {
+		log.Println("errors in license classification ", errs)
+	}
+
+	// extract results
+	results := c.classifier.GetResults()
+	sort.Sort(results)
+
+	// process results to update the normalizedLicenseMap
+	if results.Len() > 0 {
+		for _, r := range results {
+			if licenseVal, ok := filesLicenseMap.Get(r.Filename); ok {
+				// since results are sorted, we can skip processing of data with confidence <90%
+				if r.Confidence < 0.9 {
+					break
+				}
+
+				// skip processing since a higher confidence result is already processed
+				if licenseVal.ClassificationConfidence > r.Confidence {
+					continue
+				}
+
+				licenseVal.ClassificationConfidence = r.Confidence
+				filesLicenseMap.Set(r.Filename, licenseVal)
+
+				// update normalized license map
+				normalizedLicenseMap[licenseVal.LicenseKey] = r.Name
+			}
+		}
+	}
+
+	defer func() {
+		// update normalized license map for license keys which couldn't be classified or had no url in pom for classification
+		uniqLicenseKeys := c.uniqueLicenseKeys.Items()
+		for key, license := range uniqLicenseKeys {
+			if _, ok := normalizedLicenseMap[key]; !ok {
+				normalizedLicenseMap[key] = license.Name
+			}
+		}
+
+		err := fileutil.WriteJSON(c.licensedir+types.NormalizedlicenseFileName, normalizedLicenseMap)
+		if err != nil {
+			log.Println(err)
+		}
+	}()
+
+	return nil
+}
+
+func (c *Crawler) prepareClassifierData() cmap.ConcurrentMap[string, License] {
```

Comment:
I looked on build action. This function worked 1:25 hours. 
Can we update this logic to up work speed?

What if we will use similar logic as in `Crawl` function?
We can use semafore, use limit of gorutines from options, etc...
We can save urls in channel and take them in loop.

wdyt?

Reply:
I'm suggesting this because I'm worried about maintaining this code. It will be hard to find a bug if you have to wait 1.5 hours for 1 Crawler launch.

Reply:
Yup. Have been trying to figure out why this function takes so much time. Ideally license file text shouldn't be huge. Downloading them and writing contents to files concurrently shouldn't ideally take so much time. Wondering if i am missing some nuance. 
My guess is , it's because of the use of concurrent maps (uniqueLicenseKeys and filesLicenseMap) within the function. They are thread safe so blocking each other.

> What if we will use similar logic as in Crawl function?
> We can use semafore, use limit of gorutines from options, etc...
> We can save urls in channel and take them in loop.

 Are you suggesting to use a similar framework? Not sure if this is going to make a difference.

Reply:
I can avoid `filesLicenseMap`. It need not be a concurrent map. There are other ways to achieve it.

Regarding `uniqueLicenseKeys`, we can extract the contents into a normal map, since we just need a get. 

Let me give that a try

Reply:
>  Not sure if this is going to make a difference.

If i understand correctly - you use 10 parallel goroutines to write license files. Can we increase this value to 1000?

Reply:
Yeah i meant, number of go routines won't make a difference unless i get rid of the concurrent thread safe maps being used. Let me figure out a workaround and get back to you.

Reply:
Combination of concurrent maps and retry-able http client config , seemed to be the problem.

Fixed both. 
- Not using the concurrent map within the goroutine for file generation .
- Using a simple http client with timeout
- Have restructured the goroutine handling as well . Kept it same as crawl

Have triggered the action. Let's see if it is better now.

Reply:
action returns error(((

Reply:
Yup. Http client timeout error is not being handled. Updated it

Reply:
Seems to be fine now. Takes 31m for crawl

Reply:
yep. It is really good. 
Let's wait result size of db.

Reply:
new sizes:
454.36 MB is size before unpacking (Trivy will download image with this size).
792 MB - is db size in cache dir.

current sizes:
452.25 MB is size before unpacking.
743M - is db size.

LGTM.
Good job, @namandf !

Reply:
Great :)